[FLINK-32481][table] Support type inference for procedure

luoyuxia opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
To support type inference for procedure


## Brief change log
  -  Consider the procedure as a special function in internal implementation since they are almost simlar in the concept and it make the procedure can share the infra has built for function
  -  Extract some common logic from `FunctionMappingExtractor` used for function type inference to `BaseMappingExtractor` so that it can be shared by `ProcedureMappingExtractor` for type inference of procure
  -  Create `ProcedureMappingExtractor` which extends `BaseMappingExtractor` to infer type for procedure as well as some adjustment for procedure


## Verifying this change
Added test in `TypeInferenceExtractorTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

luoyuxia commented:
@flinkbot run azure

luoyuxia commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java:
```diff
@@ -143,6 +145,27 @@ public static TypeInference forAsyncTableFunction(
+                        ProcedureMappingExtractor.createParameterSignatureExtraction(1),
+                        ProcedureMappingExtractor.createReturnTypeResultExtraction(),
+                        ProcedureMappingExtractor.createParameterAndReturnTypeVerification());
+        try {
+            return extractTypeInferenceOrError(mappingExtractor);
+        } catch (Throwable t) {
+            throw extractionError(
+                    t,
+                    "Could not extract a valid type inference for procedure class '%s'. "
+                            + "Please check for implementation mistakes and/or provide a corresponding hint.",
+                    mappingExtractor.getFunctionClass().getName());
+        }
+    }
```
Good suggestion.

luoyuxia replied:
Good suggestion. But it may be not as so good to extract a common method considering we also want to throw the exception that it's an error in type inference for procedure or function. If a common method is extracted, we may need to use if/else or another flag to mark it which I think is not as so good.
But I would like to extract this snippet to a method to follow the style of type inference for function.

LadyForest commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java:
```diff
@@ -133,10 +149,19 @@ private static class DefaultAnnotationHelper {
         // no implementation
     }
 
+    @ProcedureHint
```
Maybe we can simplify the changes like 
```java
Index: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java b/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java
--- a/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java	(revision 0dae399acf5aaf04cdc3cfddada4a7b38d7f2395)
+++ b/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java	(date 1689241074609)
@@ -26,6 +26,7 @@
 
 import javax.annotation.Nullable;
 
+import java.lang.annotation.Annotation;
 import java.util.Arrays;
 import java.util.Objects;
 import java.util.function.Function;
@@ -42,8 +43,6 @@
 @Internal
 final class FunctionTemplate {
 
-    private static final FunctionHint DEFAULT_ANNOTATION = getDefaultAnnotation();
-
     private final @Nullable FunctionSignatureTemplate signatureTemplate;
 
     private final @Nullable FunctionResultTemplate accumulatorTemplate;
@@ -144,35 +143,27 @@
 
     // --------------------------------------------------------------------------------------------
 
+    @ProcedureHint
     @FunctionHint
     private static class DefaultAnnotationHelper {
         // no implementation
     }
 
-    @ProcedureHint
-    private static class DefaultProcedureAnnotationHelper {
-        // no implementation
-    }
-
-    private static FunctionHint getDefaultAnnotation() {
-        return DefaultAnnotationHelper.class.getAnnotation(FunctionHint.class);
-    }
-
-    private static ProcedureHint getDefaultProcedureHintAnnotation() {
-        return DefaultProcedureAnnotationHelper.class.getAnnotation(ProcedureHint.class);
+    private static <H extends Annotation> H getDefaultAnnotation(Class<H> hClass) {
+        return DefaultAnnotationHelper.class.getAnnotation(hClass);
     }
 
     private static <T> T defaultAsNull(FunctionHint hint, Function<FunctionHint, T> accessor) {
-        final T defaultValue = accessor.apply(DEFAULT_ANNOTATION);
-        final T actualValue = accessor.apply(hint);
-        if (Objects.deepEquals(defaultValue, actualValue)) {
-            return null;
-        }
-        return actualValue;
+        return defaultAsNull(hint, getDefaultAnnotation(FunctionHint.class), accessor);
     }
 
     private static <T> T defaultAsNull(ProcedureHint hint, Function<ProcedureHint, T> accessor) {
-        final T defaultValue = accessor.apply(getDefaultProcedureHintAnnotation());
+        return defaultAsNull(hint, getDefaultAnnotation(ProcedureHint.class), accessor);
+    }
+
+    private static <T, H extends Annotation> T defaultAsNull(
+            H hint, H defaultHint, Function<H, T> accessor) {
+        final T defaultValue = accessor.apply(defaultHint);
         final T actualValue = accessor.apply(hint);
         if (Objects.deepEquals(defaultValue, actualValue)) {
             return null;

```

luoyuxia replied:
Thanks for the sugeestion. I like it.

LadyForest commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TemplateUtils.java:
```diff
@@ -69,6 +85,21 @@ static Set<FunctionTemplate> asFunctionTemplates(
+                            try {
+                                return FunctionTemplate.fromAnnotation(typeFactory, hint);
+                            } catch (Throwable t) {
+                                throw extractionError(t, "Error in function hint annotation.");
```
Nit: Error in procedure hint annotation

LadyForest reviewed (commented):
Sorry for the delay. LGTM in general and I only left some minor comments.

luoyuxia commented:
@LadyForest Thanks for your reviewing. I have addressed your comment in this [commit](https://github.com/apache/flink/pull/22904/commits/35a62a159cd1fa45aaa250218f013cd43d8d31d1).

luoyuxia commented:
@LadyForest Thanks for reviewing. Merging...
